### PR TITLE
Deploy openshift-apiserver and kube-apiserver

### LINKED ISSUE #3636: Deploy openshift 4 apiserver

### PR DESCRIPTION
**Pull request by alvaroaleman**
**What this PR does / why we need it**:

The remaining pieces to deploy the openshift four and openshift four kube-apiserver

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3636 

**Special notes for your reviewer**:

**Documentation**:

**Does this PR introduce a user-facing change?**:

```release-note
none
```

### CODE REVIEW

File: api/pkg/controller/openshift/resources/controlplane_configmap.go [hunk 1]
```diff
@@ -31,22 +32,54 @@ type masterConfigData interface {
...
+
+func OpenshiftAPIServerConfigMapCreator(data openshiftAPIServerCreatorData) reconciling.NamedConfigMapCreatorGetter {
 	return func() (string, reconciling.ConfigMapCreator) {
 		return openshiftAPIServerConfigMapName, func(cm *corev1.ConfigMap) (*corev1.ConfigMap, error) {
 			if cm.Data == nil {
 				cm.Data = map[string]string{}
 			}
-			masterConfig, err := getMasterConfig(ctx, data, "apiserver")
+			apiServerConfigBuffer := bytes.Buffer{}
+			tmpl, err := template.New("base").Funcs(sprig.TxtFuncMap()).Parse(openshiftAPIServerConfigTemplate)
```
Comment (mrIncompetent) [inline api/pkg/controller/openshift/resources/controlplane_configmap.go]: Something i noticed during the AWS changes & found quite nice:
```go
var (
    controlPlanePolicyTpl = template.Must(template.New("worker-policy").Parse(`my-awesome-template-content`))
)
```
That way we fail very early when the template is broken. 
Though not a request for change here - Maybe something for future improvements

Comment (alvaroaleman) [reply api/pkg/controller/openshift/resources/controlplane_configmap.go]: woah that is awesome

File: api/pkg/controller/openshift/resources/controlplane_configmap.go [hunk 2]
```diff
@@ -31,22 +32,54 @@ type masterConfigData interface {
...
+func OpenshiftKubeAPIServerConfigMapCreator(data masterConfigData) reconciling.NamedConfigMapCreatorGetter {
+	return func() (string, reconciling.ConfigMapCreator) {
+		return openshiftKubeAPIServerConfigMapName, func(cm *corev1.ConfigMap) (*corev1.ConfigMap, error) {
+			if cm.Data == nil {
+				cm.Data = map[string]string{}
+			}
+			apiServerConfigBuffer := bytes.Buffer{}
+			tmpl, err := template.New("base").Funcs(sprig.TxtFuncMap()).Parse(openshiftKubeAPIServerConfigTemplate)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse config template for openshift Kube apiserverr: %v", err)
```
Comment (mrIncompetent) [inline api/pkg/controller/openshift/resources/controlplane_configmap.go]: `apiserverr` -> `apiserver`

File: api/pkg/controller/openshift/resources/controlplane_configmap.go [hunk 3]
```diff
@@ -152,6 +185,219 @@ spec:
...
+
+const openshiftKubeAPIServerConfigTemplate = `admission:
+  pluginConfig:
+    network.openshift.io/RestrictedEndpointsAdmission:
+      configuration:
+        apiVersion: network.openshift.io/v1
+        kind: RestrictedEndpointsAdmissionConfig
+        restrictedCIDRs:
+        - 10.128.0.0/14
+        - 172.30.0.0/16
```
Comment (mrIncompetent) [inline api/pkg/controller/openshift/resources/controlplane_configmap.go]: Are those the service & pod CIDR's ?
If so, we might want to template those

File: api/pkg/controller/openshift/resources/controlplane_configmap.go [hunk 4]
```diff
@@ -152,6 +185,219 @@ spec:
...
+  ca: /etc/kubernetes/pki/ca/ca.crt
+  certFile: /etc/kubernetes/kubelet/kubelet-client.crt
+  keyFile: /etc/kubernetes/kubelet/kubelet-client.key
+  port: 10250
+projectConfig:
+  defaultNodeSelector: ''
+serviceAccountPublicKeyFiles:
+- /etc/kubernetes/service-account-key/sa.pub
+servicesNodePortRange: 30000-32767
+servicesSubnet: 172.30.0.0/16
```
Comment (mrIncompetent) [inline api/pkg/controller/openshift/resources/controlplane_configmap.go]:
```yaml
servicesNodePortRange: 30000-32767
servicesSubnet: 172.30.0.0/16
```
Looks worth to replace those with the cluster specific settings

File: api/pkg/controller/openshift/resources/openshift_apiserver_deployment.go [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+				return nil, err
+			}
+			dep.Spec.Template.Labels = podLabels
+
+			return dep, nil
+		}
+	}
+}
+
+func openshiftAPIServerImage(openshiftVersion string) (string, error) {
```
Comment (mrIncompetent) [inline api/pkg/controller/openshift/resources/openshift_apiserver_deployment.go:288]: We have this now in 2 places, can we extract it?

File: api/pkg/controller/openshift/resources/openshift_apiserver_deployment.go [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+
+			etcdEndpoints := etcd.GetClientEndpoints(data.Cluster().Status.NamespaceName)
+			dep.Spec.Template.Spec.InitContainers = []corev1.Container{
+				{
+					Name:  "etcd-running",
+					Image: data.ImageRegistry(resources.RegistryGCR) + "/etcd-development/etcd:" + etcd.ImageTag,
+					Command: []string{
+						"/bin/sh",
+						"-ec",
+						fmt.Sprintf("until ETCDCTL_API=3 /usr/local/bin/etcdctl --cacert=/etc/etcd/pki/client/ca.crt --cert=/etc/etcd/pki/client/apiserver-etcd-client.crt --key=/etc/etcd/pki/client/apiserver-etcd-client.key --dial-timeout=2s --endpoints='%s' endpoint health; do echo waiting for etcd; sleep 2; done;", strings.Join(etcdEndpoints, ",")),
```
Comment (mrIncompetent) [inline api/pkg/controller/openshift/resources/openshift_apiserver_deployment.go]: This looks like a copy of the kubernetes apiserver code, can you extract it or add a todo?

File: api/pkg/controller/openshift/resources/openshift_apiserver_deployment.go (hunk 5 above)
Comment (mrIncompetent) [reply api/pkg/controller/openshift/resources/openshift_apiserver_deployment.go:288]: Todo would be fine as well

File: api/pkg/controller/usercluster/resources/openshift/namespace.go [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+			ns.Annotations = map[string]string{}
+		}
+		if ns.Labels == nil {
+			ns.Labels = map[string]string{}
+		}
+
+		ns.Annotations["openshift.io/node-selector"] = ""
+		ns.Annotations["openshift.io/sa.scc.mcs"] = "s0:c20,c15"
+		ns.Annotations["openshift.io/sa.scc.supplemental-groups"] = "1000410000/10000"
+		ns.Annotations["openshift.io/sa.scc.uid-range"] = "1000410000/10000"
```
Comment (mrIncompetent) [inline api/pkg/controller/usercluster/resources/openshift/namespace.go:21]: Can you add a comment linking to the origin?

File: api/pkg/controller/openshift/resources/openshift_apiserver_deployment.go (hunk 5 above)
Comment (alvaroaleman) [reply api/pkg/controller/openshift/resources/openshift_apiserver_deployment.go:288]: The one is for the openshift apiserver image and the other one for the openshift _kube_ apiserver image. Currently they are apparently but I don't know if that will always be the case

File: api/pkg/controller/usercluster/resources/openshift/namespace.go (hunk 7 above)
Comment (alvaroaleman) [reply api/pkg/controller/usercluster/resources/openshift/namespace.go:21]: The origin is a running openshfit cluster :grimacing:

**Comment by alvaroaleman** [comment]
@mrIncompetent adressed everything except templating of `servicesNodePortRange`, as we don't allow configuring that

**Comment by mrIncompetent** [comment]
> adressed everything except templating of servicesNodePortRange, as we don't allow configuring that

We do because i screwed something up in the past & a setup using this flag is still alive.

https://github.com/kubermatic/kubermatic/blob/master/api/cmd/kubermatic-controller-manager/options.go#L95

**Comment by mrIncompetent** [comment]
Oh but since this is a new type of cluster, we might not make it configurable ?

**Comment by alvaroaleman** [comment]
@mrIncompetent that is for the seed, not for the usercluster or do I remember that wrong?

**Comment by mrIncompetent** [comment]
We can indeed omit the NodePort range flag here, because the Kube apiserver for OpenShift does not use the `--kubernetes-service-node-port` flag.

For Kubernetes clusters we have to make the NodePort range configurable - because of the usage of `--kubernetes-service-node-port` (It requires that the listening port is within the NodePort range). 
Though that something we can potentially remove in the future.

**Pull request merged**